cocos/renderer/CCVertexIndexBuffer.h

```diff
@@ -25,94 +26,326 @@
     
-    GLuint getVBO() const;
+    // @brief increases the capacity of the buffer by count elements
+    //        optionally zeroes out the elements.
+    void addCapacity(size_t count, bool zero = false);
     
-protected:
-    VertexBuffer();
-    virtual ~VertexBuffer();
+    // @brief swaps elements in the buffer without resizing the buffer
+    //        if defer is true, then the native buffer is not updated.
+    void swapElements(size_t source, size_t dest, size_t count);
+
+    // @brief moves elements in the buffer to the dest index.
+    //        if defer is true, then the native buffer is not updated.
+    void moveElements(size_t source, size_t dest, size_t count);
+
+    // @brief if dirty, copies elements to the client buffer (if any)
+    // and optionally submits the elements to the native buffer (if any)
+    // if elements is null, then the entire client is commited to native. 
+    void bindAndCommit(const void* elements = nullptr, size_t count = 0, size_t begin = 0);
+
+    size_t getSize() const
+    {
+        return getElementCount() * getElementSize();
+    }
+    
+    size_t getCapacityInBytes() const
+    {
+        return getCapacity() * getElementSize();
+    }
+    
+    CC_DEPRECATED_ATTRIBUTE uint32_t getVBO() const
+    {
+        return _vbo;
+    }
+
+    void setElementCount(size_t count)
+    {
+        CCASSERT(count <= _capacity, "element count cannot exceed capacity");
+        if (count != _elementCount)
+        {
+            _elementCount = count;
+            setDirty(true);
+        }
+    }
+    
+    size_t getElementCount() const
+    {
+        return _elementCount;
+    }
+    
+    size_t getElementSize() const
+    {
+        return _elementSize;
+    }
+    
+    size_t getCapacity() const
+    {
+        return _capacity;
+    }
+    
+    bool hasClient() const
+    {
+        return _arrayType & ArrayType::Client;
+    }
+    
+    bool hasNative() const
+    {
+        return _arrayType & ArrayType::Native;
+    }
+    
+    bool isDirty() const
+    {
+        return _dirty;
+    }
+    
+    void setDirty(bool dirty)
+    {
+        _dirty = dirty;
+    }
+    
+    void clear();
+
+    void recreate() const;
+    
+    // @brief returns the client array if present, otherwise nullptr
+    template <typename T = void>
+    T* getElementsT() const
+    {
+        return hasClient() ? static_cast<T*>(_elements) : nullptr;
+    }
+
+    // @brief append
+    template <typename T>
+    size_t appendElementsT(T* source, size_t elements = 1, bool defer = true)
+    {
+        CCASSERT(0 == sizeof(T) % getElementSize(), "elements must divide evenly into elementSize");
+        auto mult = sizeof(T) / getElementSize();
+        appendElements((const void*)source, mult * elements, defer);
+        return getSize();
+    }
+    
+    template <typename T>
+    void updateElementsT(const T* element, size_t count, size_t begin, bool defer = true)
```

Comment:
Do we need the post-fix `T`? Because other template functions don't have the post-fix.